### PR TITLE
Backend: DelayedRun sync

### PR DESCRIPTION
**Pull request by Thunderblade73**
Found a better sync bug fix.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt [hunk 1]
```diff
@@ -1,34 +1,24 @@
...
 
-import at.hannibal2.skyhanni.utils.LorenzUtils.editCopy
+import java.util.concurrent.LinkedBlockingQueue
 import kotlin.time.Duration
 
-// TODO find better sync bug fix than creating a new map for each use
 object DelayedRun {
-    var map = mapOf<() -> Any, SimpleTimeMark>()
+    val map = mutableListOf<Pair<() -> Any, SimpleTimeMark>>()
+    private val inMap = LinkedBlockingQueue<Pair<() -> Any, SimpleTimeMark>>()
```
Comment (lineargraph) [inline src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt]: You should not be using a `LinkedBlockingQueue` here. We don't want any blocking behaviour. Use  `ConcurrentLinkedQueue` instead.

File: src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt [hunk 2]
```diff
@@ -1,34 +1,24 @@
 package at.hannibal2.skyhanni.utils
 
-import at.hannibal2.skyhanni.utils.LorenzUtils.editCopy
+import java.util.concurrent.LinkedBlockingQueue
 import kotlin.time.Duration
 
-// TODO find better sync bug fix than creating a new map for each use
 object DelayedRun {
-    var map = mapOf<() -> Any, SimpleTimeMark>()
+    val map = mutableListOf<Pair<() -> Any, SimpleTimeMark>>()
```
Comment (lineargraph) [inline src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt]: The naming of both of these variables could use some improvement: Maybe `tasks` and `delayedTasks`.

File: src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt [hunk 3]
```diff
@@ -1,34 +1,24 @@
...
-            entries.removeIf { (runnable, time) ->
-                val inPast = time.isInPast()
-                if (inPast) {
-                    runnable()
-                }
-                inPast
-            }
+        inMap.drainTo(map)
+        map.removeIf { (runnable, time) ->
+            time.isInPast().also { if (it) runnable() }
```
Comment (lineargraph) [inline src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt]: I don't think the `also` one line style here serves to improve readability.

**Review by lineargraph — CHANGES_REQUESTED** [review]

**Review by lineargraph — APPROVED** [review]

**Review by lineargraph — APPROVED** [review]
Just had a thought about using a TreeMap instead :upside_down_face:, but no, this looks great.

File: src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt [hunk 4]
```diff
@@ -1,34 +1,29 @@
...
-                }
-                inPast
+        tasks.removeIf { (runnable, time) ->
+            val inPast = time.isInPast()
+            if (inPast) {
+                runnable()
             }
+            inPast
         }
+        futureTasks.drainTo(tasks)
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt:28]: I think drainTo should happen before removeIf

File: src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt [hunk 5]
```diff
@@ -1,34 +1,29 @@
...
+    private val futureTasks = ConcurrentLinkedQueue<Pair<() -> Any, SimpleTimeMark>>()
 
     fun runDelayed(duration: Duration, run: () -> Unit) {
-        map = map.editCopy {
-            this[run] = SimpleTimeMark.now() + duration
-        }
+        futureTasks.add(Pair(run, SimpleTimeMark.now() + duration))
     }
 
     fun runNextTick(run: () -> Unit) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/utils/DelayedRun.kt:16]: The current logic feels like it is "run in 2 ticks from now", right?

**Review by hannibal002 — CHANGES_REQUESTED** [review]
The logic looks fancy, thanks <3
Not sure if im missing something but one small thing here looks misleading or wrong to me

**Comment by Thunderblade73** [comment]
Since both suggestion are for one and the same issue:
It's 1 tick must happend but up to 1 extra tick could have happend.
I wanted to have this behavior so it is guaranteed that one tick has parsed, without that we would always execute the runNextTick in the subTick before the next Tick which is misleading. 
And also some stupid stuff could happen without that:
Say you run a runNextTick in a LorenzTickEvent. Then it could be possible that the code executes immediately after that event Listener finishes since the next Listener could be the checkRuns one and therefore no time has parsed for it.

**Comment by hannibal002** [comment]
The idea behind "run in next tick" for me was not a "provide at least 50ms delay" but rather "run in the next tick in the main thread".
But your logic is better.
Now the function executes in between 50 and 100 ms, instead of 1 and 50 ms. this feels more consistent.
And the extra 50ms will hurt no one.

Now only one question remains to me: should we rename the function runNextTick to something else?

**Comment by Thunderblade73** [comment]
The name is fine. Also documented the behavior.

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**